dynamic_instrumentation/src/instrument.rs

```diff
@@ -62,14 +65,22 @@ impl Instrumenter {
     }
 
     /// Finish instrumentation and write out metadata to `metadata_file_path`.
-    pub fn finalize(&self, metadata_file_path: &Path) -> anyhow::Result<()> {
+    pub fn finalize(&self, metadata_path: &Path) -> anyhow::Result<()> {
         let mut locs = self.mir_locs.lock().unwrap();
         let mut functions = self.functions.lock().unwrap();
         let locs = locs.drain(..).collect::<Vec<_>>();
         let functions = functions.drain().collect::<HashMap<_, _>>();
         let metadata = Metadata { locs, functions };
         let bytes = bincode::serialize(&metadata).context("Location serialization failed")?;
-        fs_err::write(metadata_file_path, &bytes).context("Could not open metadata file")?;
+        let mut file = OpenOptions::new()
+            .append(true)
+            .write(true)
+            .open(metadata_path)
+            .context("Could not open metadata file")?;
+        file.file().lock_exclusive()?;
+        let e = file.write_all(&bytes);
+        file.file().unlock()?;
+        e?;
```

Comment:
I wasn't sure how exactly to do this.  Am I correct in not returning the error until after unlocking the file so that we don't leave the file locked if there is a write error?  And is there a reason there's not a lock guard or something similar returned from locking?

Reply:
I looks right to me to unlock before returning the error. I'd be curious if `write_all` could panic even though it returns a result. My guess about the locking guard is to match what the underlying sys crates are doing.

Reply:
`write_all` shouldn't panic unless something is seriously wrong with the `write` syscall.